tools/halide_image.h

```diff
@@ -1,278 +1,668 @@
+        return (dimensions > 1) ? dim[1].extent : 1;
+    }
+    int channels() const {
+        return (dimensions > 2) ? dim[2].extent : 1;
+    }
+    // @}
+
+    /** Make a new image which is a deep copy of this image. Use crop
+     * or slice followed by copy to make a copy of only a portion of
+     * the image. The new image uses the same memory layout as the
+     * original, with holes compacted away. */
+    Image<T, D> copy() const {
+        Image<T, D> src = *this;
+
+        // Reorder the dimensions of src to have strides in increasing order
+        int swaps[(D*(D+1))/2];
+        int swaps_idx = 0;
+        for (int i = dimensions-1; i > 0; i--) {
+            for (int j = i; j > 0; j--) {
+                if (src.dim[j-1].stride > src.dim[j].stride) {
+                    std::swap(src.dim[j-1], src.dim[j]);
+                    swaps[swaps_idx++] = j;
+                }
             }
         }
-    }
 
-    Image &operator=(const Image &other) {
-        Contents *p = other.contents;
-        if (p) {
-            p->ref_count++;
+        // Make a copy of it using this dimension ordering
+        Image<T, D> dst = src;
+        dst.allocate();
+
+        // Concatenate dense inner dimensions into contiguous memcpy tasks
+        Image<T, D> src_slice = src;
+        Image<T, D> dst_slice = dst;
+        int64_t slice_size = 1;
+        while (src_slice.dimensions && src_slice.dim[0].stride == slice_size) {
+            assert(dst_slice.dim[0].stride == slice_size);
+            slice_size *= src_slice.dim[0].stride;
+            src_slice = src_slice.sliced(0, src_slice.dim[0].min);
+            dst_slice = dst_slice.sliced(0, dst_slice.dim[0].min);
         }
-        if (contents) {
-            contents->ref_count--;
-            if (contents->ref_count == 0) {
-                delete contents;
-                contents = NULL;
-            }
+
+        slice_size *= sizeof(T);
+        // Do the memcpys
+        src_slice.for_each_element([&](const int *pos) {
+                memcpy(dst_slice.address_of(pos), src_slice.address_of(pos), slice_size);
+            });
+
+        // Undo the dimension reordering
+        while (swaps_idx > 0) {
+            int j = swaps[--swaps_idx];
+            std::swap(dst.dim[j-1], dst.dim[j]);
         }
-        contents = p;
-        return *this;
-    }
 
-    T *data() { return (T*)contents->buf.host; }
+        return dst;
+    }
 
-    const T *data() const { return (T*)contents->buf.host; }
+    /** Make an image that refers to a sub-range of this image along
+     * the given dimension. */
+    Image<T, D> cropped(int d, int min, int extent) const {
+        // Make a fresh copy of the underlying buffer (but not a fresh
+        // copy of the allocation, if there is one).
+        Image<T, D> im = *this;
+        im.dim[d].min = min;
+        im.dim[d].extent = extent;
+        int shift = min - dim[d].min;
+        im.host += shift * dim[d].stride * sizeof(T);
```

Comment:
I think this should assert that the dev field is 0. Same for anything else that mutates the host pointer.